### ISSUE TITLE
count=0 and splats issues

### ISSUE THREAD

**Issue by gozer** [hashicorp/terraform#6889]
Terraform v0.7.0-dev (8cc00d6848bdd531a6fd656ab2d372d51de6a075)

I think it's simpler to illustrate with just my example.

```
variable "count1" { }

variable "count2" { }

resource "aws_eip" "test1" {
  count = "${var.count1}"
}

resource "aws_eip" "test2" {
  count = "${var.count2}"
}

output "EIPs" {
  value = "${compact(concat(element(aws_eip.test1.*.id,0),element(aws_eip.test2.*.id,0)))}"
}
```
```bash
$> terraform apply -var 'count1=1' -var 'count2=1'
[...]
Outputs:

EIPs = [
  eipalloc-dfdba7bb
  eipalloc-82dfa3e6
]
```

```bash
$> terraform apply -var 'count1=1' -var 'count2=0'
aws_eip.test2: Refreshing state... (ID: eipalloc-82dfa3e6)
aws_eip.test1: Refreshing state... (ID: eipalloc-dfdba7bb)
aws_eip.test2: Destroying...
aws_eip.test2: Destruction complete

Apply complete! Resources: 0 added, 0 changed, 1 destroyed.
```
Notice the absence of any outputs? Feels like a bug, and it's bitten me in more than once place with optionnal resources I turn off with the count=0 trick.

It tends to infect different resources with this kind of vanishing output phenomenom.

**Comment by apparentlymart** [comment]
@gozer am I understanding correctly here that you were expecting something like this?

```
Outputs:

EIPs = []
```

**Comment by gozer** [comment]
No, if you look at the example, should have been:


EIPs = [
  eipalloc-dfdba7bb
]

**Comment by gozer** [comment]
It looks like the fact that one resource having been disabled with count=0 infects the whole output and makes it just vanish somehow...

**Comment by jen20** [comment]
Hi @gozer! This should have been fixed by 0.7rc2 - there was an issue around handling splat resources with a count of zero or one, and outputs currently swallow interpolation error messages (due to complexity with destroy) - but can be seen using TF_LOG=INFO (I believe). If this is _not_ fixed by RC2 can you please reopen the issue? Thanks!

**Comment by thecastor** [comment]
Hey guys,

I don't think this bug has been fully resolved yet.

**I'm using Terraform 0.7.9**

### Test Configuration

```
variable "cnt" {
  default = 0
}

resource "null_resource" "test1" {
  count = "${var.cnt}"
}

resource "null_resource" "test2" {
  count = 1
}

output "test_out" {
  value = ["${ concat(null_resource.test1.*.id, null_resource.test2.*.id})"]
}
```
### To reproduce
1. ```terraform plan && terraform apply```. _As expected_, `test_out=["<ID2>"]`.
2. Set `cnt=1`, then `terraform plan && terraform apply`. `test_out=["<ID1>","<ID2>"]`, _as expected_.
3. Set `cnt=0`, then `terraform plan && terraform apply`. The behaviour is that described by @gozer. `null_resource.test1` is removed (_as expected_), and the _complete_ `output` _disappears_. It is no longer in the state file. The same behaviour shows for, e.g., a `null_resource` instead of an `output`.
4. `terraform apply` again (no changes to the configuration). _The expected result appears_. `test_out=["<ID2>"]`
